### ISSUE TITLE
BackstopJS x Symfony x login pages

### ISSUE THREAD

**Issue by DaPo** [garris/BackstopJS#1141]
Hi @garris and thank you for your awesome job on your awesome creation.

I may have an issue with BackstopJS. Not an issue per say but I think I need some advices.

Indeed I'm working on this website that is running with Symfony and I want to test some pages that need to be logged in to be visited. Like user's parameters or so.

So I tried with cookies on a `cookies.json` file but cookies are moving with PHP so it's not a solution for me.

I was looking around and saw [this gist](https://gist.github.com/torounit/33cdc09cbb698f4b4287085da23c7fe5) to get logged in via HTTP Authentification (saw [this](https://github.com/garris/BackstopJS/issues/123) and [this](https://github.com/garris/BackstopJS/issues/153#issuecomment-179366960). So I switched for Chromy instead of Pupet:

```json
"report": ["browser", "CI"],
  "engine": "chromy",
  "engineOptions": {
    "chromeFlags": "/Applications/GoogleChrome.app",
    "args": ["--no-sandbox"]
  },
```

When I lunch the tests I have the following output:

```
 ./node_modules/backstopjs/cli/index.js test --filter="Header - Anonymous"
BackstopJS v4.4.2
Loading config:  /Users/cyril/workspace/myawesomewebsite/website/backstop.json

COMMAND | Executing core for "test"
createBitmaps | Selected 1 of 11 scenarios.
searching for 5 available ports.
These ports will be used: [9222,9223,9224,9225,9226]
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=1333,1610"],"port":9222,"waitTimeout":30000,"visible":false,"chromePath":"/Applications/GoogleChrome.app","args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=1210,1610"],"port":9223,"waitTimeout":30000,"visible":false,"chromePath":"/Applications/GoogleChrome.app","args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=992,2270"],"port":9224,"waitTimeout":30000,"visible":false,"chromePath":"/Applications/GoogleChrome.app","args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=768,1840"],"port":9225,"waitTimeout":30000,"visible":false,"chromePath":"/Applications/GoogleChrome.app","args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=320,1840"],"port":9226,"waitTimeout":30000,"visible":false,"chromePath":"/Applications/GoogleChrome.app","args":["--no-sandbox"]}
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
(node:32499) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Uncaught exception: spawn /Applications/GoogleChrome.app EACCES Error: spawn /Applications/GoogleChrome.app EACCES
    at Process.ChildProcess._handle.onexit (internal/child_process.js:266:19)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
/Users/cyril/workspace/myawesomewebsite/website/node_modules/backstopjs/cli/index.js:46
      throw err;
      ^

Error: spawn /Applications/GoogleChrome.app EACCES
    at Process.ChildProcess._handle.onexit (internal/child_process.js:266:19)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
Emitted 'error' event on ChildProcess instance at:
    at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
    at onErrorNT (internal/child_process.js:468:16)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  errno: -13,
  code: 'EACCES',
  syscall: 'spawn /Applications/GoogleChrome.app',
  path: '/Applications/GoogleChrome.app',
  spawnargs: [
    '--disable-translate',
    '--disable-extensions',
    '--disable-background-networking',
    '--safebrowsing-disable-auto-update',
    '--disable-sync',
    '--metrics-recording-only',
    '--disable-default-apps',
    '--mute-audio',
    '--no-first-run',
    '--remote-debugging-port=9222',
    '--user-data-dir=/var/folders/p2/kqp21z716rb6sbl827sg3v5w0000gn/T/lighthouse.XXXXXXX.lp9ECBXQ',
    '--disable-background-networking',
    '--disable-background-timer-throttling',
    '--disable-client-side-phishing-detection',
    '--disable-default-apps',
    '--disable-hang-monitor',
    '--disable-popup-blocking',
    '--disable-prompt-on-repost',
    '--disable-sync',
    '--enable-automation',
    '--enable-devtools-experiments',
    '--metrics-recording-only',
    '--no-first-run',
    '--password-store=basic',
    '--safebrowsing-disable-auto-update',
    '--use-mock-keychain',
    '--headless',
    '--hide-scrollbars',
    '--mute-audio',
    '--disable-gpu',
    '--force-device-scale-factor=1',
    '--disable-infobars=true',
    '--window-size=1333,1610',
    'about:blank'
  ]
}
```

So I tried to remove that `chromePath` thing and:
```
$ ./node_modules/backstopjs/cli/index.js test --filter="Header - Anonymous"
BackstopJS v4.4.2
Loading config:  /Users/cyril/workspace/myawesomewebsite/website/backstop.json

COMMAND | Executing core for "test"
createBitmaps | Selected 1 of 11 scenarios.
searching for 5 available ports.
These ports will be used: [9222,9223,9224,9225,9226]
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=1333,1610"],"port":9222,"waitTimeout":30000,"visible":false,"args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=1210,1610"],"port":9223,"waitTimeout":30000,"visible":false,"args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=992,2270"],"port":9224,"waitTimeout":30000,"visible":false,"args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=768,1840"],"port":9225,"waitTimeout":30000,"visible":false,"args":["--no-sandbox"]}
Starting Chromy: {"chromeFlags":["--disable-gpu","--force-device-scale-factor=1","--disable-infobars=true","--window-size=320,1840"],"port":9226,"waitTimeout":30000,"visible":false,"args":["--no-sandbox"]}
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
Cookie state restored with: []
(node:32552) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
      COMMAND | Executing core for "report"
      compare | Chromy error: Error. See scenario – Header - Anonymous (1333+)
      compare | Chromy error: Error. See scenario – Header - Anonymous (1210 → 1332)
      compare | Chromy error: Error. See scenario – Header - Anonymous (992 → 1331)
      compare | Chromy error: Error. See scenario – Header - Anonymous (768 → 991)
      compare | Chromy error: Error. See scenario – Header - Anonymous (320 → 767)
       report | Test completed...
       report | 0 Passed
       report | 5 Failed
       report | Writing jUnit Report
       report | jUnit report written to: backstopjs/backstop_data/ci_report/xunit.xml
       report | Writing browser report
       report | Resources copied
       report | Failed json report copy to: /Users/cyril/workspace/myawesomewebsite/website/backstopjs/backstop_data/bitmaps_test/20200124-122541/report.json
       report | Copied jsonp report to: /Users/cyril/workspace/myawesomewebsite/website/backstopjs/backstop_data/html_report/config.js
      COMMAND | Executing core for "openReport"
   openReport | Attempting to ping
   openReport | Remote not found. Opening backstopjs/backstop_data/html_report/index.html
       report | *** Mismatch errors found ***
      COMMAND | Command "report" ended with an error after [0.08s]
      COMMAND | Error: Mismatch errors found.
                    at /Users/cyril/workspace/myawesomewebsite/website/node_modules/backstopjs/core/command/report.js:189:17
                    at processTicksAndRejections (internal/process/task_queues.js:97:5)
      COMMAND | Command "test" ended with an error after [1.874s]
      COMMAND | Error: Mismatch errors found.
                    at /Users/cyril/workspace/myawesomewebsite/website/node_modules/backstopjs/core/command/report.js:189:17
                    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

So here I am, I don't know where to look further more. If you have any advices I'd love to read about them.

Thank you for your time.

Cheers!

**Comment by garris** [comment]
I recommend using puppeteer.

You should absolutely use cookies, If they change often then you may need to create a separate backstop job which logs in, and saves cookies once before running your tests.  Search for puppeteer log-in scripts on google.  Then use that info to create a custom job prior to running tests.  Scenarios run in parallel so it must be a separate job.

**Comment by DaPo** [comment]
Okay, I'll look into that. Thank you for your answer and thank you again for BackstopJS.

Cheers!

**Comment by DaPo** [comment]
Hi, sorry to bother you again but I can,'t find any solution.

I manage to reload my cookies.json with puppeter, but I can't tell to BackstopJS to test the website with the cookies. I copied my files [on gist](https://gist.github.com/DaPo/93a0640f2d7ab23837e12053fecb81e5) can you take a look at them?

Here is what's happening:
- I launch backstop with `$ node_modules/backstopjs/cli/index.js test --filter="Header - user2" --verbose`
- Chromium is opening /connexion
- it fills the field with user2 / user2
- the backstopjs' report is opening in my Firefox, the only thing I see in the "test" part is a screenshot where I see I'm not logged
- Chromium clic on the "connect" button then I see on Chromium my homepage, logged.

Do you have any clue on what I'm doing wrong here? :( 

Thank you for your time.

**Comment by garris** [comment]
I do not think this will work...  
https://gist.github.com/DaPo/93a0640f2d7ab23837e12053fecb81e5#file-loadcookies-js-L6

Try manually editing your cookies.json file first with the correct cookies.

**Comment by DaPo** [comment]
Okay, I think I have something wrong with the loadcookies.js indeed.

At this time I have `loadCookies.js` + `onBefore.js` that are the same as the default ones from your repo. In `cookies.json` I have the cookies I got with my old `loadCookies.js`and it works.

So now I know my cookies are working, how can I scrap them before each scenario from my website as I wanted to do so with my gist files?

**Comment by DaPo** [comment]
So I managed to make things work, I updated [the gist](https://gist.github.com/DaPo/93a0640f2d7ab23837e12053fecb81e5).

I have a problem tho, when the `cookies.json` doesn't exists, the `JSON.parse(fs.readFileSync(cookiePath, 'utf8'));` from loadCookies returns an error even if cookies.json is created. I just have to relaunch the command to make it works so that's OK for me I guess.

Thank you again for your time and your help. ✌️
